### DISCUSSION TITLE
Move dampening logic into PhantomCameraHost

### DISCUSSION THREAD

**Discussion by ZenithStar** [ramokz/phantom-camera#278] [Ideas]
With the jitter introduced in moving from _physics_process to _process, I believe there needs to be a refactor moving dampening logic out of PhantomCamera2D/PhantomCamera3D and into PhantomCameraHost. This is where the tweening logic already is. Dampening parameters can still be associated with each PhantomCamera and read in by PhantomCameraHost in the same way that PhantomCameraTween parameters are currently. PhantomCamera nodes should act as the "setpoint" or the "model" that can be updated asynchronously or during a timestep of the user's choosing, similar to [AnimationMixer.callback_mode_process](https://docs.godotengine.org/en/stable/classes/class_animationmixer.html#class-animationmixer-property-callback-mode-process), while Camera3D/Camera2D nodes should act as the "view" and always update in the _process/IDLE timestep. 

In addition, all dampening parameters should always be exposed, possibly as a resource similar to PhantomCameraTween and with multiple dampening schemes. In the current system, the typical use-case for FollowMode.THIRD_PERSON doesn't allow for LookAt dampening to be used, because orientation is being set by user input, rather than LookAtMode logic. As for multiple dampening schemes, `class_name PhantomCameraDampener3D extends Resource` can provide an interface like `func _dampen(input: Transform3D, delta: float) -> Transform3D` and individual dampener implementations `class_name PhantomCameraSmoothDamp3D extends PhantomCameraDampener3D` can override this interface. (2D and 3D dampener logic will need to be separated.)

**Comment by ZenithStar** [comment]
In addition, the same goes for when implementing the camera shake feature. Offsets resulting from an impulse should only affect the change in `transform` for the Camera3D+PhantomCameraHost. The PhantomCameras should only contain the data for the impulse(s), possibly as `class_name PhantomCameraImpulse extends RefCounted`, with its `transform` unaffected.

**Comment by ramokz** [comment]
Jitter will gradually be mitigated as Godot introduces physics interpolation to 2D and 3D — only 2D is close to being [officially supported](https://godotengine.org/article/dev-snapshot-godot-4-3-dev-6/). There's a longer discussion about this in #241 and a PR for 2D Physics interpolation in #294, which is pending changes in the dev branch for Godot 4.3 before being released. The current implementation is a necessary compromise until Godot implements those two feature variants, as it doesn't impede non-physics animations from being jittery.

The system was originally damping the movement in the `PCamHost` script, i.e. applying it directly to the `Camera2D/3D`, but was changed for a couple of reasons:

1. If you apply the damping to the `Camera` node directly, it can lead to tweens behaving in a weird, mechanical and, likely, undesirable manner.
Say if the `Camera`, with a `PCamHost`, is currently attached to and follows a `PCam`. The `Camera` then updates its position based on its assigned `PCam`'s position and then applies damping on itself on top of it — in other words, it's dragging behind the `PCam`. It then becomes a question of where should the camera tween to if the next `PCam` has a moving `follow target` with `follow damping` applied. If a tween gets triggered, it would first stop following its current target, like how it currently works, and then transition towards the newly active `PCam`, which is where the complication appears. The `Camera` would move towards where that new active `PCam` is based on its tween duration and transition properties. But if that new `PCam` has `following damping` set, then the tween's target is set to be precisely where it's `follow target`, and `follow offset` if applied, is because the damping isn't factored into the `PCam`'s position. The issue with that is the moment the tween finishes, the camera would then start dragging behind, rather than tween towards a target position that already is dragging behind its target, i.e. the dampened position, to begin with. To me, that felt like a less seamless and more awkward outcome. It can't really apply both a tween with a specific duration _and_ have damping applied at the same time, as the latter would interfere with the speed / animation ease of the tween. So it would result in a fairly abrupt transition from a tween finishing to then applying damping if applied directly to the `Camera` node.

2. For `Third Person` follow, the issue I see with having the damping applied to the `Camera`, or `PCam3D` for that matter, instead of the `SpringArm3D` is that it means updating the rotation would automatically apply damping without the user explicitly declaring that. Since if you rotate the `SpringArm3D` by using one of the `Third Person` rotation methods, it will also change the `PCam`'s position, with it being a child of the `SpringArm3D`. Because the position changes, it will then apply damping to the camera, which means that the user has less control over how the rotation behaves as `follow damping` now directly interferes with it.

3. To the point about the shake/noise, there is a [work-in-progress branch](https://github.com/ramokz/phantom-camera/tree/camera-noise) that has an early, but functional, implementation of this mechanic, which does apply the camera noise/shake to the `Camera2D`/`Camera3D` directly. The approach here is done for the same reasons you mention. Like with the first point, if damping was happening on the camera directly, it would mean that the shake has to effectively ignore the damping if both needed to occur at the same time. For `Camera2D`, this is less of an issue since it uses the `Offset` property, making it not reliant on any positional changes, but creating camera noise for the `Camera3D` is a bit more complicated. Like with its 2D counterpart, it has a `V/H Offset`, but conversely the shake doesn't need to just apply a positional change, but rather positional _and_ rotational. The positional changes also need to support being applied in the z-axis, which rules out relying on the `V/H Offset` properties, as they don't support any changes in depth. Combine this behaviour with damping, like with the point about tween, it would essentially need to disable the damping since it would otherwise work on top of the noise pattern.

**Reply by ZenithStar** [reply]
The holistic way to describe any dampening and tweening (as well as any noise function) is a [transfer function](https://en.wikipedia.org/wiki/Transfer_function) between a model x(t) and the view y(t). This is far from intuitive to your typical user, and so you created the far more easy-to-understand interface of tweening ease, trans, and duration parameters, as well as dampening parameters that specify just duration per axis. However, this is a hack for the sake of user friendliness and ease of implementation due to the availability of the nice transition functions provided by Godot Tween. In fact, using Tween at all is a hack. Take for example this situation:

![image](https://github.com/ramokz/phantom-camera/assets/12200619/0f37dd03-f821-4d5c-ad4e-9dae9a59666c)


When I wrote my own camera interface before I started using PhantomCamera, I described this transfer function using a [PID controller](https://en.wikipedia.org/wiki/Proportional%E2%80%93integral%E2%80%93derivative_controller). (This is code inside the Camera3D script, and `setpoint` would be the equivalent of a PhantomCamera.)
```GDScript
@export_group("Camera Easing", "camera_ease_")
@export var camera_ease_P: float = 10.0
@export var camera_ease_I: float = 0.01
@export var camera_ease_I_clamp: float = 0.5
@export var camera_ease_D: float = -20.0
var _last_error = Vector3.ZERO
var _integral_error = Vector3.ZERO
@export var rotational_ease_duration: float = 0.1
func _physics_process(delta):
	var error = (setpoint.global_position - global_position) * delta
	_integral_error = clamp(_integral_error + error, Vector3(-1.0,-1.0,-1.0)*camera_ease_I_clamp, Vector3(1.0,1.0,1.0)*camera_ease_I_clamp)
	global_position += error * camera_ease_P + _integral_error * camera_ease_I + (error - _last_error) * camera_ease_D
	_last_error = error
	quaternion = quaternion.slerp(setpoint.quaternion, delta / rotational_ease_duration)
```

In my opinion, [CSS transitions](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Transitions/Using_CSS_transitions) is the ideal form that this interface should take. The user simply specifies transition (i.e. transfer function) parameters, which can be chosen from a preset list like linear to ease-in-out, just like Godot Tweens, to a [user-specified cubic bezier function](https://developer.mozilla.org/en-US/docs/Web/CSS/transition-timing-function). After specifying this transfer function, users can asynchronous set DOM values and the renderer acts as a layer of abstraction between that model and the rendered view. Godot does not have built-in support for this layer of abstraction, and so it must be user-created by projects like this and [GuiTransitions](https://github.com/murikistudio/simple-gui-transitions). Using GuiTransitions as an example, in order to support some common CSS effects like sliding in and out, that project needed to utilize [a shader](https://github.com/murikistudio/simple-gui-transitions/blob/godot-4/addons/simple-gui-transitions/shaders/transform.gdshader) to create this layer of abstraction and disconnect the Godot "DOM" model from the rendered view.

Another way to think about it is that PhantomCamera should act as a [retained mode](https://en.wikipedia.org/wiki/Retained_mode) interface to Camera3D+PhantomCameraHost's [immediate mode renderer interface](https://en.wikipedia.org/wiki/Immediate_mode_GUI). In the current implementation, PhantomCamera must be treated as an immediate mode interface when dampening is active, otherwise the jitter occurs. However, it can be a retained mode interface when inactive or when tweening is active. Moving dampening back into PhantomCameraHost allows PhantomCamera to be a retained mode interface in all situations, which is strongly preferable.

For your comments regarding third person follow mode, updating the rotation on the spring arm *should* trigger positional dampening because the position of the camera *is* changing. The user is already explicitly specifying that they want this dampening to happen when they set a non-zero positional dampening duration.

**Reply by ZenithStar** [reply]
![image](https://github.com/ramokz/phantom-camera/assets/12200619/cb21c913-1fd7-4e08-a238-8dd2dc9847bc)

**Reply by ramokz** [reply]
If I'm reading it right, your point about tweening is that if another `PCam` becomes active that is moving at a high velocity, then the `Camera` will be tweened to the position of where that `PCam` is at the time the tween begin. But **not** where it is at during the tween, let alone at the end of it?

The way it currently interpolates to a new `PCam` during a tween is that it updates and calculates the target position every frame. So if the `PCam` that is being transitioned to moves during a tween, then so too will its destination. So the blue curved line in your illustration is how it behaves at the minute. If the targeted `PCam` moves at a high velocity, then it will increase the tweened `Camera`'s velocity to fulfil the duration promise. The end position of a tweened `Camera` will always be the position of the newly active `PCam` during the final tween animation frame. The only compromise of this is that the easing will be stretched to satisfy the duration parameter. But that is unavoidable given a predetermined Bézier-curve cannot create a consistent interpolation algorithm if the distance it has to travel to changes mid-flight while respecting the tween's duration.

The common usage of Godot's Tween is indeed a static `to` and `from` property interpolation — i.e. what the red line shows. The way it's being used in the addon, however, is a more manual and flexible approach that the class offers called [`_interpolate_value`](https://docs.godotengine.org/en/stable/classes/class_tween.html#class-tween-method-interpolate-value). In this context of the tweening system used here, it is what allows for updating the desired end state of a tween after every frame, should the target change.

So I am not entirely sure what problem the proposal is trying to solve that isn't happening already?

For the `Third Person` `follow mode` rotation changes, I still don't think that `follow damping` should affect the `Camera` if you apply a rotational value with the setter methods — for three reasons.
For one, it means that if you apply _any_ amount of `follow damping`, then the user can never achieve fully précising orbiting controls — it will always have a minimum damping value applied to whatever the user applies. I.e. it takes control away from the user.
The second reason is that it forces the user to set a `follow damping value` that influences both the movement of the `PCam` when the target moves _and_ the rotation around it when the user programmatically chooses to change it. Meaning if the user wants to follow a target with some degree of damping, but wishes for a slightly less dampened motion when orbiting its target, then that can't be done.
Thirdly, it breaks the general expected behaviour of how it works in other `follow modes`. The damping is only ever applied if the target's position changes, whereas with applying `follow damping` to the orbit rotation as well means that the damping is no longer `follow damping` but rather `position changed damping`. It also feels counterintuitive if what is externally and internally marked and written as “follow damping” would ever get triggered when the `follow target` isn't moving.

Think if the user wants to dampen the rotation of the camera around its `follow target`, then it's up to them to interpolate the values they supply the rotational setters with. It keeps the rotational behaviour unopinionated and fully customizable based on what the user wants.

**Reply by ZenithStar** [reply]
Since that's the way the tweening is implemented right now, the issue instead manifests when the transition is the other way around, when a camera transitions out from a fast moving PhantomCamera. The ideal way to transition between separate PhantomCameras, rather than directly tweening the parameters of the Camera3D+PhantomCameraHost towards the new PhantomCamera's parameters, the Camera3D+PhantomCameraHost's parameter values should be a cross-fade composite between the parameters of the two PhantomCameras e.g. `Camera3D.global_transform = old_pcam.global_transform.interpolate_with( new_pcam.global_transform, tween_this_value_instead )`

Regardless, the primary goal of this discussion is to decouple the rate at which the PhantomCamera node processes its transform from the render frame rate. The stutter/jitter is just a symptom of this underlying design flaw and will manifest in other ways. For example, when you get around to #26, how will you resolve multiple cameras in multiple windows on multiple monitors with multiple frame rates?

The process of writing this reply has lead me to boot up Unity for the first time in forever to see [how Cinemachine does it](https://docs.unity3d.com/Packages/com.unity.cinemachine@3.1/manual/CinemachineBrain.html). Cinemachine does not even provide an option for updating inside Update/_process. The default is "Smart Update" which dynamically chooses the same update interval as the follow target's, either in FixedUpdate/_physics_process or LateUpdate/call_deferred. The default fallback update for when blending between two different virtual cameras is LateUpdate/call_deferred.

As for the third person follow dampening, the whole point is that the user is not allowed to make any fully precise movement when damping is enabled. Cinemachine acts the way I described. Take for example a simple test scene with Body = "3rd Person Follow" and Aim = "Same As Follow Target" and a high aim damping. If I try to rotate the camera constantly to the right at a faster rate than the damping allows the camera to able to keep up, eventually the camera will start rotating left to catch up, because a left turn becomes the shortest path to achieve the desired orientation setpoint. This demonstrates that at some layer of abstraction there is a transform setpoint/model value that is decoupled from that of the resultant camera positions'. For Cinemachine, this transform appears to be an internal private variable inside CinemachineVirtualCamera. This probably works for them because Unity does not have a built-in SpringArm node that uses the scene tree hierarchy for its transform. One possible solution for PhantomCamera can be to also forsake the use of a SpringArm3D node and instead simulate one in software using a ShapeCast3D.

**Reply by ZenithStar** [reply]
A summary of how Cinemachine works
![image](https://github.com/ramokz/phantom-camera/assets/12200619/268d276e-bd62-4adc-b7ed-8c80f2c41eb8)

**Reply by ramokz** [reply]
So I want to point back to this part of a previous message:
> Jitter will gradually be mitigated as Godot introduces physics interpolation to 2D and 3D — only 2D is close to being [officially supported](https://godotengine.org/article/dev-snapshot-godot-4-3-dev-6/). There's a longer discussion about this in https://github.com/ramokz/phantom-camera/issues/241 and a PR for 2D Physics interpolation in https://github.com/ramokz/phantom-camera/pull/294, which is pending changes in the dev branch for Godot 4.3 before being released. The current implementation is a necessary compromise until Godot implements those two feature variants, as it doesn't impede non-physics animations from being jittery.

The above is precisely about supporting both `_process` and `_physics_process` logic for the `PCam` system without the user having to do any extra work. Can even find a commented out code block inside the PR's `phantom_camera_host.gd` file that mentions [`interpolation_mode`](https://github.com/ramokz/phantom-camera/blob/259a49c7845668fc42757139ea032e65188b178c/addons/phantom_camera/scripts/phantom_camera_host/phantom_camera_host.gd#L36), which is what Unity refers to as `Update Method` in CinemachineBrain. So what you're referring to is not only known, but is very much planned and already, effectively, done — at least for 2D.

It's worth repeating that the reason the addon doesn't do this already is because Godot doesn't have built-in support for physics interpolation. Unity is a much more mature engine, whereas Godot is still catching up, where this is one of the areas that's only just now is starting to become part of the core engine. The only way to support it at the minute is to do it manually, which is why both the [doc site](https://phantom-camera.dev/support/faq#option-1-recommended-%E2%80%94-the-smoothing-addon) but also [Godot themselves](https://docs.godotengine.org/en/stable/tutorials/rendering/jitter_stutter.html#jitter) refer to the [`smoothing-addon`](https://github.com/lawnjelly/smoothing-addon/tree/4.x), as that is effectively what that engine should be doing natively but just currently doesn't. Worth mentioning that the author behind the smoothing-addon is also the one who implemented it for Godot 3, and someone has since then started porting it to Godot 4 — starting with Godot 4.3 for 2D. Do fully agree that it's a fundamental flaw with the addon, but this is ultimately a limitation with the current state of the engine that has no simple solution.

---

For the `Third Person` `follow mode`, I get what you mean, and I can see the value in what you're describing.
The `SpringArm3D` was chosen as it mirrors how other systems uses a camera for third-person behaviour — Unreal being a good example. They have a separate property for damping the rotation of the spring arm called [Camera Rotation Lag](https://dev.epicgames.com/documentation/en-us/unreal-engine/using-spring-arm-components-in-unreal-engine), "lag" referring to what we call "damping", as a separate system to the `follow damping`, in Unreal called `Camera Lag`. Now, I'm not saying Unreal = right, Cinemachine = wrong, They have different methods to similar systems and both are valid. What I'm ultimately interested in is finding a solution that solves people's desired outcomes. What you're raising here is a valid point about its limitation, and so that makes for a question about how to go about it, which I think warrants further discussion. 

There are areas I would personally like to see improvements in, such as having a smoothed camera collision, rather than instantaneously jumping to the spot between a collider and the `follow target`. Something the `SpringArm3D` cannot support in its current form. But while I think Godot's implementation of `SpringArm3D` isn't perfect, it gave the most mileage with the least amount of effort to get it to the functional state that it's currently in — while also leveraging any improvements `SpringArm3D` might get. Going beyond that is something I would be interested in, but it would take a degree of effort that would come at the cost of other features that I want to support first before optimizing existing, and arguably functional, ones.

**Reply by ZenithStar** [reply]
Regardless of whether or not interpolation becomes engine built-in or if it's in another add-on, **it is a hack** to glue together the retained mode interface that the combinations of user input and the output of a physics engine and the immediate mode interface that is mandated by the way GPUs and computer monitor hardware works. While it is an unavoidable hack somewhere along this pipeline, by depending on this hack, you have turned `follow_target` into an immediate mode interface.  This is counter to all the other features in this package, which are asynchronous retained mode interfaces. This package does not need to be and *should not* be dependent on such a hack.

Is there a particular limitation keeping you from wanting to do this? Having reviewed the way Cinemachine works, I think it would be fine to clone its behavior one-to-one. This would mean, rather than my original proposal to move the damping logic from PhantomCamera into PhantomCameraHost, PhantomCameras are instead all converted to `top_level` in all modes and its transforms are always manually assigned by script, rather than by the scene tree.

**Reply by ramokz** [reply]
To be honest, I still don't quite get what the fundamental issue is. I'm not a graphic programmer nor someone with in-depth knowledge of game physics, but I would like to understand what you're trying to explain so that I can see where you're coming from. The state of the way the addon interpolates isn't ideal, I agree. Running it with the changes coming in 4.3, and with the PR mentioned earlier, addresses the core issue I've had since forever with physics framerate affecting the camera movement and switching between moving targets that operate in `_process` and `_physics_process`.
What I would like to understand is, what is the practical issue yet to be resolved from that future change?

> Is there a particular limitation keeping you from wanting to do this?

In short, time. If an existing feature already fulfils the basics needs, where a more complex overhaul would be desired to satisfy other outcomes, then that becomes a case of managing time vs. the gains of said overhaul. It's very difficult to justify spending a significant amount of time improving a particular `follow mode` when there are other, more fundamental, improvements to be made.

It's also worth pointing out that nothing prevents someone from making their own system on top of the base addon. The `follow modes` are ultimately shortcuts to behavioural solutions. One can always make their own follow system and simply make the `PCam` do what they want in a separate script, much like a normal camera, and keep the other parts of the system. Obviously, the intent of the project is to strip away complexity with setting that up, but equally I think it's difficult to support every possible need. Unlike Cinemachime, this is purely maintained as an outside-of-daily-work project.

**Reply by ZenithStar** [reply]
I just took a look at how the 3.6 engine physics interpolation is implemented, and it is indeed a far better solution than the smoothing addon. The smoothing add-on is a non-solution as illustrated here:
![image](https://github.com/ramokz/phantom-camera/assets/12200619/c9c0b2ad-63db-492d-a797-2a62214f8fad)

>It's also worth pointing out that nothing prevents someone from making their own system on top of the base addon.

This is the problem right now. All my custom code interfacing to the SpringArm3D motion is tied to the physics rate. Mouse and gamepad inputs are polled at 125Hz which are buffered and read by the 60Hz physics, while my main monitor refreshes at 144Hz. Basically, the solution from my side would need to be to write in 125Hz -> dynamic frame rate interpolation code in order to apply rotations to the spring arm.

With the 3.6/4.3 physics interpolation active, all phantom camera code, including both PhantomCameraHost and PhantomCamera2D/3D would be required to operate in _physics_process. That's fine for when that time comes, but for now, my solution has to be to run on a 60FPS monitor or to change 
https://github.com/ramokz/phantom-camera/blob/d8a6e248dcc1d26f01c8c73bc07105daf81507fd/addons/phantom_camera/scripts/phantom_camera/phantom_camera_3d.gd#L548
back to `func _physics_process(delta: float) -> void:`

**Reply by ramokz** [reply]
Have you had a chance to read the posts in #241?
It discusses the `_process` vs. `_physics_process` and what the direction will be.

**Reply by ZenithStar** [reply]
In the case of 3rd person follow mode, SpringArm3D is inherently a physics object requiring a physics collision check that must happen during a physics tick. In 3rd person follow mode, the PhantomCamera3D becomes a non-top-level child of that physics object, and therefore also must move during the physics tick.

AFAIK, 3D physics interpolation is not implemented in 4.3: only 3.6, and who knows if it'll make it in, since 4.3 is already long overdue, and they're regularly talking about "the 4.3 feature freeze".(https://github.com/godotengine/godot/issues/89846) In general, I agree that the "fix" for any jitter from PhantomCamera's side can simply be to move everything into `_physics_process` (or at least have that option available): https://github.com/ramokz/phantom-camera/issues/241#issuecomment-2091289535

By the way @ramokz, for any of the changes I suggest either here or any of the issues, I wasn't directly asking if you could implement them, as I have some version, if not multiple versions of most of these changes in one project or another. I just wasn't sure if it would be worth my time to format them as clean and modular PRs, since many changes are codependent and require larger-scale refactors.

**Reply by ramokz** [reply]
> AFAIK, 3D physics interpolation is not implemented in 4.3: only 3.6, and who knows if it'll make it in, since 4.3 is already long overdue, and they're regularly talking about "the 4.3 feature freeze".(https://github.com/godotengine/godot/issues/89846) In general, I agree that the "fix" for any jitter from PhantomCamera's side can simply be to move everything into _physics_process (or at least have that option available): https://github.com/ramokz/phantom-camera/issues/241#issuecomment-2091289535

3D will likely not make it to 4.3, as 2D barely did, but hard to say.

Want to point to [this particular comment](https://github.com/ramokz/phantom-camera/issues/241#issuecomment-2085115749) in that post for why moving the logic into the `_physic_process` isn't a good idea before the physics interpolation gets added to Godot.
Tldr; you solve one problem (jitter free physics-based nodes) but introduce another (jittery non-physics-based nodes). If you add the PCam logic to the `_physics_process` you will not be able to have any jitter-free animations for anything moving using a Godot `Tween`, `AnimationPlayer` or any non-physics-based animation. This is fine when you're _only_ working with an object that's moving in the `_physics_process`, but is pretty bad in all other scenarios. Think cutscenes, animated in-world UI etc.
Conversely, the [guidelines in the documentation site](https://phantom-camera.dev/support/faq#i-m-seeing-jitter-what-can-i-do) explain how to get around addressing both situations, so you can have jitter-free animations in both `_physics_process` and `_process`.

Again, that workaround is to use the `smoothing-addon`, or part of its underlying code. The author is also the one who added it as a built-in feature to Godot 3. The reason being is that it's something that should be in the engine core, and the  `smoothing-addon` was initially made an external solution to fill in a gap the engine had. Where they later on then decided to make that addon effectively obsolete by making the feature built-in instead. What audeck says in the link you're referring to is correct in the context of having physics interpolation as part of the engine itself. But without it, it resurfaces the issue mentioned above. Phantom Camera _will_ dynamically switch between `_process` and `_physics_process`, and you can see the current setup for 2D scenes in #294, depending on the assigned target for when physics interpolation is available. However, it's being limited to 2D and for Godot 4.3 and later for the moment for the same reasons mentioned above.

> By the way @ramokz, for any of the changes I suggest either here or any of the issues, I wasn't directly asking if you could implement them, as I have some version, if not multiple versions of most of these changes in one project or another. I just wasn't sure if it would be worth my time to format them as clean and modular PRs, since many changes are codependent and require larger-scale refactors.

I wouldn't suggest making any PRs without bringing up either a discussion or feature issue post about the exact feature or changes proposed — as per the [contribution documentation page](https://phantom-camera.dev/support/how-to-contribute#%E2%8C%A8%EF%B8%8F-addon-contributions). Larger changes need some degree of alignment about the direction and what problem a given addition would solve, or whether if the proposal even has merit. Some proposals are sound for a given project, but can be so bespoke to a particular thing that it adds more noise than benefit to the project as a whole. Like pretty much all things for all projects, features ought to be planned and agreed upon before work starts, otherwise time and energy is likely just wasted.
